File: Week_03/buildTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        hashmap= {v:i for i,v in enumerate(inorder)}
        def buildTreeHelper(preorder,p_start,p_end,inorder,i_start,i_end):
            if p_start==p_end:
                return None
            root=TreeNode(preorder[p_start])
            i_index=hashmap[root.val]
            LEFT=i_index-i_start
            root.left=buildTreeHelper(preorder,p_start+1,p_start+LEFT+1,inorder,i_start,i_index)
            root.right=buildTreeHelper(preorder,p_start+LEFT+1,p_end,inorder,i_index+1,i_end)
            return root
        return buildTreeHelper(preorder,0,len(preorder),inorder,0,len(inorder))

File: Week_03/buildTree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
    hashmap:=map[int]int{}
    for i,v :=range inorder{
        hashmap[v]=i
    }
    var buildTreeHelper func(perorder []int, p_start int, p_end int, inorder []int, i_start int, i_end int) *TreeNode
    buildTreeHelper = func(perorder []int, p_start int, p_end int, inorder []int, i_start int, i_end int) *TreeNode {
        if p_start==p_end{
            return nil
        }
        root_val:=preorder[p_start]
        root := &TreeNode{root_val, nil, nil}
        i_index:=hashmap[root_val]
        LEFT:=i_index-i_start
        root.Left=buildTreeHelper(preorder,p_start+1,p_start+LEFT+1,inorder,i_start,i_index)
        root.Right=buildTreeHelper(preorder,p_start+LEFT+1,p_end,inorder,i_index+1,i_end)
        return root
    }
    return buildTreeHelper(preorder,0,len(preorder),inorder,0,len(inorder))
}
